### PR TITLE
ssh-server-in-go

### PR DESCRIPTION
**Pull request by Unknwon**

### CODE REVIEW

**Comment by mattfarina** [comment]
Can you please move the post from the `upcoming` directory to the `content/advent-2015` directory.

File: upcoming/ssh-server-in-go.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
++++
+author = ["Jiahua Chen"]
+date = "2015-12-18T04:07:33-05:00"
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: This date is before an already published post. Can you update the date to the December 19th.

File: upcoming/ssh-server-in-go.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+title = "Writing SSH server in Go"
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: Consider putting _an_ in front of _SSH_ so it reads more clearly. _"Writing an SSH server in Go"_

File: upcoming/ssh-server-in-go.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+series = ["Advent 2015"]
++++
+
+When I'm working on [Gogs](https://gogs.io) project, there is a need of builtin
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: Considering adding _the_ to _When I'm working on [Gogs](https://gogs.io) project_ so it reads _When I'm working on **the** [Gogs](https://gogs.io) project_

File: upcoming/ssh-server-in-go.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+SSH server, which allows users to preform Git-only operations through key-based
+authentication.
+
+All available resources on the web are all minimal examples and does not fit this
+specific requirement. Therefore, I think it's worth sharing my experiences to make
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: > All available resources on the web are all minimal examples and does not fit this specific requirement.

would read more easily as something like:

> The available resources on the web are all minimal examples and do not fit this specific requirement.

File: upcoming/ssh-server-in-go.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+your life easier in case you just ran into same problem as mine.
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: Consider changing _just ran_ into _run_.

File: upcoming/ssh-server-in-go.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+The code structure is pretty much same to the examples you can find on the web.
+
+1. Start a SSH listening host.
+2. Accept new requests and validate their public key with database.
+3. Preform Git operations.
+4. The most important part, return a status if no error occurs.
+
+OK, before we get started, just note that code are not supposed to be copy-paste
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: > just note that code are not supposed to

would read more cleanly as

> just note that the code examples are not supposed to

File: upcoming/ssh-server-in-go.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+directory because users will add this key to their `known_hosts` file.
+
+```go
+// Listen starts a SSH server listens on given port.
+func Listen(port int) {
+	config := &ssh.ServerConfig{
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			pkey, err := models.SearchPublicKeyByContent(strings.TrimSpace(string(ssh.MarshalAuthorizedKey(key))))
+			if err != nil {
+                // handle error
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: The indentation on this is spaces while the rest of the code is tabs. Can you change this to tabs. The difference causes display issues on the blog.

File: upcoming/ssh-server-in-go.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+func listen(config *ssh.ServerConfig, port int) {
+	listener, err := net.Listen("tcp", "0.0.0.0:"+com.ToStr(port))
+	if err != nil {
+		panic(err)
+	}
+	for {
+		// Once a ServerConfig has been configured, connections can be accepted.
+		conn, err := listener.Accept()
+		if err != nil {
+            // handle error
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: The indentation on this is spaces while the rest of the code is tabs. Can you change this to tabs. The difference causes display issues on the blog.

File: upcoming/ssh-server-in-go.md [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+			continue
+		}
+		// Before use, a handshake must be performed on the incoming net.Conn.
+		sConn, chans, reqs, err := ssh.NewServerConn(conn, config)
+		if err != nil {
+            // handle error
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: The indentation on this is spaces while the rest of the code is tabs. Can you change this to tabs. The difference causes display issues on the blog.

File: upcoming/ssh-server-in-go.md [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+		}
+
+		// The incoming Request channel must be serviced.
+		go ssh.DiscardRequests(reqs)
+		go handleServerConn(sConn.Permissions.Extensions["key-id"], chans)
+	}
+}
+```
+
+1. Accept requests inside a infinite `for` loop.
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: > a infinite

would read more clearly as:

> an infinite

File: upcoming/ssh-server-in-go.md [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+func handleServerConn(keyID string, chans <-chan ssh.NewChannel) {
+	for newChan := range chans {
+		if newChan.ChannelType() != "session" {
+			newChan.Reject(ssh.UnknownChannelType, "unknown channel type")
+			continue
+		}
+
+		ch, reqs, err := newChan.Accept()
+		if err != nil {
+            // handle error
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: The indentation on this is spaces while the rest of the code is tabs. Can you change this to tabs. The difference causes display issues on the blog.

File: upcoming/ssh-server-in-go.md [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+goroutines so the connection won't be blocked.
+
+Finally, we're getting into the most interesting part.
+
+1. There could be more than one request from single channel, we need to handle each
+of them.
+2. The payload comes from request somehow is not always in a clean format, so we
+have to preform a clean operation to remove unless characters:
+
+    ```go
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: Can you remove the spaces before _```go_. This is causing display problems on the blog.

File: upcoming/ssh-server-in-go.md [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+func cleanCommand(cmd string) string {
+	i := strings.Index(cmd, "git")
+	if i == -1 {
+		return cmd
+	}
+	return cmd[i:]
+}
+    ```
```
Comment (mattfarina) [inline upcoming/ssh-server-in-go.md]: Can you remove the spaces before _```_. This is causing display problems on the blog.

**Comment by mattfarina** [comment]
I enjoyed reading this post.Thank you. There are a couple things that need to be fixed to display on the blog and I made a few grammar suggestions to make it easier to read.

File: upcoming/ssh-server-in-go.md (hunk 7 above)
Comment (Unknwon) [inline upcoming/ssh-server-in-go.md]: I did use tabs...

**Comment by Unknwon** [comment]
Thanks, updates. 

Just note Atom shows me I do use tabs in the code blocks... not sure how GitHub handles it.

**Comment by mattfarina** [comment]
@Unknwon thank you for the post. It's now live at https://blog.gopheracademy.com/advent-2015/ssh-server-in-go/

**Comment by Unknwon** [comment]
Thank you!

**Pull request merged**